db/fetcher/encoded_doc.go

```diff
@@ -211,15 +234,35 @@ func (encdoc *encodedDocument) Decode() (*client.Document, error) {
 
 // DecodeToDoc returns a decoded document as a
 // map of field/value pairs
-func (encdoc *encodedDocument) DecodeToDoc(mapping *core.DocumentMapping) (core.Doc, error) {
-	doc := mapping.NewDoc()
-	doc.SetKey(string(encdoc.Key))
-	for fieldDesc, prop := range encdoc.Properties {
+func (encdoc *encodedDocument) DecodeToDoc() (core.Doc, error) {
+	return encdoc.decodeToDoc(false)
+}
+
+func (encdoc *encodedDocument) decodeToDocForFilter() (core.Doc, error) {
+	return encdoc.decodeToDoc(true)
+}
+
+func (encdoc *encodedDocument) decodeToDoc(filter bool) (core.Doc, error) {
+	if encdoc.mapping == nil {
```

Comment:
related to my comment on the `mapping` prop, I see the fact that we have to check this here (in the hot/usage path) and potentially error as a symptom of a bad constructor/type.

Reply:
the cost of checking this is miniscule compare to the cost of decoding and running the filter.

Same answer to my thoughts above. Fetcher now needs to run filters, filters need a `core.Doc`, which you can only get by using a `core.DocumentMapping`.

If we use `map[string]any` to run the filter on the fetcher (like we did pre-mapper days) then we can remove the dependency on the mapping stuff, but then we double the decoding work.

Reply:
I'm much less worried about performance here - I flagged this as it is an error path that we do not/should not need in the 'doing' part of the code

Reply:
If we need to use a pointer's value within the function that doesn't define it, then we need to check that it's not nil. I feel like it makes sense to do that here.

Reply:
> I'm much less worried about performance here - I flagged this as it is an error path that we do not/should not need in the 'doing' part of the code

This doesn't address my point that we *must* do this here, I see no way around this. Moreover, this is based on how you wrote the mapper/filter/doc.

Tangentially, your noted objection to these `x != nil` come up all the time in reviews, and we keep reminding you that its as command and idiomatic as it gets for Go code.

Reply:
Even if it's a function parameter it can still be nil, it doesn't change anything.

In fact, it is a function parameter, just at a different function / call site to initialize the fetcher.

It should still be checked.

--

Regarding the idiomatic. Ultimately it doesn't matter what other languages do. This *Is* Go, and Go convention wins. The language authors and core maintainers have been very explicit that the goals as a language are not to be fancy, and will often ignore more modern PL techniques and research, in an attempt at "simplicity" and clarity.

The pointer has to be checked regardless. Putting it in something like a homegrown results or option container doesn't change that, nor does making it a function argument.

My comment was directed at the historical nature of this thread, as it comes up often, and the response is often the same on both ends.

PS. Word for word my previous response was 50% directed at you lol. The first half of the message was at your suggestion, as I pointed out there's almost no way around having the doc mapper in the fetcher unless either refactor the filter stuff, or double the decoding work. Neither seem ideal compared to leaving as is.

Reply:
There's a reply here in the other comment thread you raised the issue on too, with my same points, from Friday.

https://github.com/sourcenetwork/defradb/pull/1500#discussion_r1192614605

Reply:
Resolving this convo, as it is going nowhere, and is of very low value lol.  Leave as-is if prefered.